File: java/153-Find-Minimum-in-Rotated-Sorted-Array.java
class Solution {
    public int findMin(int[] nums) {
        int l = 0;
        int r = nums.length - 1;
        
        while (l <= r) {
            if (nums[l] <= nums[r]) {
                return nums[l];
            }
            
            int mid = (l + r) / 2;
            if (nums[mid] >= nums[l]) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return 0;
    }
}

File: scala/153-Find-Minimum-In-Rotated-Sorted-Array.scala
object Solution {
    def findMin(nums: Array[Int]): Int = {
        var res = nums(0)
        var (left, right) = (0, nums.length - 1)
        
        while (left <= right) {
            if (nums(left) < nums(right)) {
                res = res.min(nums(left))
                left = right + 1 // break condition
            } else {
                val mid = (left + right) / 2
                res = res.min(nums(mid))
                
                if (nums(mid) >= nums(left)) {
                    left = mid + 1
                } else {
                    right = mid - 1
                }               
            }
        }
        
        return res
    }
}
